File: gitdot-web/packages/api/src/resource/migration.ts
import { z } from "zod";

export const GitHubInstallationResource = z.object({
  id: z.uuid(),
  installation_id: z.number(),
  owner_id: z.uuid(),
  installation_type: z.string(),
  github_login: z.string(),
  created_at: z.iso.datetime(),
});
export type GitHubInstallationResource = z.infer<
  typeof GitHubInstallationResource
>;

export const GitHubRepositoryResource = z.object({
  id: z.number(),
  name: z.string(),
  full_name: z.string(),
  description: z.string().nullable(),
  private: z.boolean(),
  default_branch: z.string(),
  pushed_at: z.iso.datetime().nullable(),
});
export type GitHubRepositoryResource = z.infer<typeof GitHubRepositoryResource>;

export const MigrationRepositoryResource = z.object({
  id: z.uuid(),
  origin_full_name: z.string(),
  destination_full_name: z.string(),
  visibility: z.string(),
  status: z.string(),
  error: z.string().nullable(),
  created_at: z.iso.datetime(),
  updated_at: z.iso.datetime(),
});
export type MigrationRepositoryResource = z.infer<
  typeof MigrationRepositoryResource
>;

export const MigrationResource = z.object({
  id: z.uuid(),
  number: z.number().int(),
  author_id: z.uuid(),
  origin_service: z.string(),
  origin: z.string(),
  origin_type: z.string(),
  destination: z.string(),
  destination_type: z.string(),
  status: z.string(),
  created_at: z.iso.datetime(),
  updated_at: z.iso.datetime(),
  repositories: z.array(MigrationRepositoryResource),
});
export type MigrationResource = z.infer<typeof MigrationResource>;

File: gitdot-api/src/resource/migration.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct GitHubAppInstallUrlResource {
    pub install_url: String,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct GitHubInstallationResource {
    pub id: Uuid,
    pub installation_id: i64,
    pub owner_id: Uuid,
    pub installation_type: String,
    pub github_login: String,
    pub created_at: DateTime<Utc>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct GitHubRepositoryResource {
    pub id: u64,
    pub name: String,
    pub full_name: String,
    pub description: Option<String>,
    pub private: bool,
    pub default_branch: String,
    pub pushed_at: Option<DateTime<Utc>>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MigrationResource {
    pub id: Uuid,
    pub number: i32,
    pub author_id: Uuid,

    pub origin_service: String,
    pub origin: String,
    pub origin_type: String,
    pub destination: String,
    pub destination_type: String,

    pub status: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,

    pub repositories: Vec<MigrationRepositoryResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MigrationRepositoryResource {
    pub id: Uuid,

    pub origin_full_name: String,
    pub destination_full_name: String,
    pub visibility: String,

    pub status: String,
    pub error: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
